articles/rules/preview/index.md

```diff
@@ -0,0 +1,312 @@
+
+Among many possibilities, Rules can be used to:
+
+* __Profile enrichment__: query for information on the user from a database/API, and add it to the user profile object.
+* Create __authorization rules__ based on complex logic (anything that can be written in JavaScript).
+* __Normalize attributes__ from different providers beyond what is provided by Auth0.
+* Reuse information from existing databases or APIs for migration scenarios.
+* Keep a __white-list of users__ and deny access based on email.
+* __Notify__ other systems through an API when a login happens in real-time.
+* Enable counters or persist other information. (For information on storing user data, see: [Metadata in Rules](/rules/metadata-in-rules).)
+* Enable __multifactor__ authentication, based on context (e.g. last login, IP address of the user, location, etc.).
+* Change tokens: Change the returned __scopes__ of the `access_token` and/or add claims to it, and to the `id_token`. At the moment you can do this for the [Password Exchange](/api-auth/grant/password#rules). This is possible because Rules run after the authentication and before the authorization.
+
+## Video: Using Rules
+Watch this video learn all about rules in just a few minutes.
+
+<%= include('../../videos/_video', { id: 'g7dy1fpwc3' }) %>
+
+## Rule Syntax
+
+A Rule is a function with the following arguments:
+
+* `user`: the user object as it comes from the identity provider (For a complete list of the user properties, see: [User Profile Structure](/user-profile/user-profile-structure)).
+
+* `context`: an object containing contextual information of the current authentication transaction, such as user's IP address, application, location. (A complete list of context properties is available here: [Context Argument Properties in Rules](/rules/context).)
+* `callback`: a function to send back the potentially modified `user` and `context` objects back to Auth0 (or an error).
+
+<div class="alert alert-info">
+Because of the async nature of Node.js, it is important to always call the <code>callback</code> function, or else the script will timeout.
+</div>
+
+## Examples
+
+<div class="alert alert-info">
+You can find more examples of common Rules on Github at <a href="https://github.com/auth0/rules">auth0/rules</a>.
+</div>
+
+To create a Rule, or try the examples below, go to [New Rule](${manage_url}/#/rules/create) in the Rule Editor on the dashboard.
+
+### *Hello World*
+
+This rule will add a `hello` attribute to all users authenticating through any provider.
+
+```js
+function (user, context, callback) {
+  user.hello = 'world';
+  console.log('===> set "hello" for ' + user.name);
+  callback(null, user, context);
+}
+```
+
+<div class="alert alert-info">
+  You can add <code>console.log</code> lines for <a href="#debugging">debugging</a> or use the <a href="/extensions/realtime-webtask-logs">Real-time Webtask Logs Extension</a>.
+</div>
+
+
+### Add roles to a user
```

Comment:
use case yes, example no. in the past, we would modify the user object because we would then merge it with the id_token and would return the whole package.
in the new implementation, we need to actually modify the id_token with what we want in it.

`context.idToken["http://sample"] = ['role1', 'role2', ... ]` <- pseudo code, need to actually try it and figure the actual grammar.